### ISSUE TITLE
[2017 ICLR Oral] On Large-Batch Training for Deep Learning: Generalization Gap and Sharp Minima

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#66]
Small-batch SGD는 다양한 장점을 가지고 있음

이론적으로 다음과 같은 properties를 가진다는 것이 보여짐

1. Convergence to minimizers of strongly-convex functions and to stationary points for non-convex functions
2. Saddle-point avoidance
3. Robustness to input data

하지만 small batch size와 iteration의 sequential 특성으로 인해 parallelization이 어려워 시간이 많이 걸림

이를 줄이는 가장 쉬운 방법은 large batch를 사용하는 것임

하지만 large batch를 사용하게 되면 generalization performance가 small batch에 비해 떨어지는 현상이 발견됨
(generalization gap)

해당 논문은 numerical result를 통해 large batch를 사용하게 되면 왜 generalization drop이 생기는지 밝히며, 이를 극복하기 위한 다양한 시도들에 대하여 설명함

## 1. Drawbacks of Large-Batch Methods

논문에서 large batch를 LB, small batch를 SB라고 명칭

### 1.1. Main Observation

Deep learning model을 LB로 학습할 때 generalization gap이 발생하는데, 이에 대해 다음과 같은 원인이 있을 수 있음

1. LB methods over-fit the model
2. LB methods are attracted to saddle points
3. LB methods lack the explorative properties of SB methods and tend to zoom-in on the minimizer closest to the initial point
4. SB and LB methods converge to qualitatively different minimizers with differing generalization properties

해당 논문은 1번과 2번의 요인이 아니며 3번과 4번 요인으로 인해 generalization gap이 발생하며 다음과 같이 정리함

<img width="995" alt="image" src="https://user-images.githubusercontent.com/49643709/167069287-6c29b5ff-829e-4e91-bda3-83d5b73e50a3.png">

### 1.2. Why Sharp Minimizer is Bad for Generalization?
<img width="1041" alt="image" src="https://user-images.githubusercontent.com/49643709/167069542-533b85fe-95ac-4ac4-8bcf-d06f16972f47.png">

Sharp minimizer면 왜 generalization ability가 떨어지는지에 대해 2가지 방식으로 설명할 수 있음

1. Sensitivity
    - Figure 1 참고
    - Flat minimizer varies slowly in a relatively large neighborhood
    - Sharp minimizer increases rapidly in a small neighborhood
    - Large sensitivity of the training function at a sharp minimizer negatively impacts the ability of the trained model to generalize on new data
2. Minimum description length (MDL) theory
    - Occam's razor와 비슷함
    - Statistical models that require fewer bits to describe (low complexity) generalize better
    - Flat minimum can be described with low precision, whereas a sharp minimum requires high precision
    - So flat minimizers tend to have better generalization performance than sharp minimizers

이외에도 bayesian, Gibbs energy 관점에서 설명한 방식도 있으니 관심있으면 논문의 reference를 보면 됨

### 2.1. Setup
<img width="990" alt="image" src="https://user-images.githubusercontent.com/49643709/167073021-d0f7104f-cca7-4618-ae14-3dae01b386a2.png">
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/49643709/167073407-13f04561-ac26-4cfb-b8ef-120508d9663d.png">
<img width="1003" alt="image" src="https://user-images.githubusercontent.com/49643709/167073565-b51128d3-265b-49aa-8cef-298370644ef2.png">

> - Final testing accuracy만 고려
> - Convergence trend가 어떻게 되는지는 고려 대상이 아님
> - Network마다의 train / test accuracy의 차이는 고려 대상이 아님

> - LB의 경우 학습 데이터셋의 10%를 batch size로, SB의 경우 256 batch size로 설정
> - 5 independent run에 대해 평균과 표준편차를 구함
> - Optimizer로 Adam, ADAGRAD, SGD, adaQN 등을 사용하였으며 결과는 모두 비슷함

> - Test dataset에 대한 성능이 특정 iteration을 기점으로 decay하지 않기에 over-fitting이나 over-training으로 인해 generalization gap이 벌어진 것이 아님을 알 수 있음
> - Overfitting을 방지하기 위한 early-stopping과 같은 heuristic한 기술들은 효과가 없음을 알 수 있음

### 2.2. Parametric Plots
<img width="1280" alt="image" src="https://user-images.githubusercontent.com/49643709/167076522-19630bed-ba99-4116-b445-dafcae3a5c48.png">

동일한 네트워크에 대해서 5 trials 중에서 SB, LB minimizer pair을 random하게 선택

Weight 간의 interpolation을 통해 train dataset, test dataset에 대한 loss function을 시각화

Plot 결과를 통해 LB minima가 SB minima에 비해 sharp함을 알 수 있음

#### 2.2.1. Linear
<img width="971" alt="image" src="https://user-images.githubusercontent.com/49643709/167074460-daa26066-15f5-4bb6-8e41-f151218907e1.png">

#### 2.2.2. Curvilinear
<img width="965" alt="image" src="https://user-images.githubusercontent.com/49643709/167074550-26879d44-81cb-4376-8acf-5ad8e0712df6.png">

### 2.3. Sharpness of Minima
<img width="1040" alt="image" src="https://user-images.githubusercontent.com/49643709/167082577-c0423d24-f030-4b23-b312-3a3a4dda2074.png">
<img width="1088" alt="image" src="https://user-images.githubusercontent.com/49643709/167082895-025619f7-5c60-476c-977b-855e732e8fe6.png">

Sharpness를 측정하기 위해서 f(x)의 laplace operator의 eigenvalues의 magnitude를 봐야하는데, 이는 cost가 너무 많이 들기에 논문에서는 다른 방식을 제안함

Minima인 지점 근처의 점들 중에서 가장 값이 큰 값을 찾아 이를 기준으로 측정
(L-BFGS-B를 이용)

SB, LB에서의 sharpness는 다음과 같음

<img width="795" alt="image" src="https://user-images.githubusercontent.com/49643709/167083479-9c5f7aa3-25a5-4ce3-a77b-7dcd0461f8a8.png">

Sharp minimizer가 모든 direction에 대해 빠르게 증가하는 cone 형태가 아닌 일부 direction에 대해서만 steep하다고 함
(steeply only along a small dimensional subspace, 5% of the whole space, on most other directions the function is relatively flat)

## 3. Success of Small-Batch Methods
<img width="1087" alt="image" src="https://user-images.githubusercontent.com/49643709/167085928-26977f08-ce35-43de-bef0-876053f0913b.png">

다양한 모델마다 generalization drop이 발생하는 batch size의 threshold가 존재함
(there is a batch size after which there is a large drop in testing accuracy, upward drift in value of the sharpness)

SB가 잘되는 이유는 gradient의 noise가 sharp minimizer를 빠져나오면서 빠져나올 수 없는 flatter minimizer로 움직이기 때문임

Batch size가 threshold를 넘어버리면 noise가 충분하지 않아 sharp minimizer로 빠지게 됨

<img width="1091" alt="image" src="https://user-images.githubusercontent.com/49643709/167087372-80f76627-b8b1-45d4-85d0-917953f9cc3d.png">

저자들은 더 깊이 이해하기 위해 warm-started experiment를 수행함 (piggybacked)

Warm-started large-batch solution : 처음에 SB로 모델 훈련하다 LB로 모델 훈련

Warm-start를 few initial epoch SB로 학습하고 LB로 학습하면 generalization improvement 없음

Warm-start를 certain few epoch SB로 학습하고 LB로 학습하면 accuracy도 증가하고 sharpness도 감소함

이가 의미하는 바는 warm-start할 때 SB method가 exploration phase를 끝내고 flat minimizer를 찾았다는 말임

강화학습 관점에서 봤을 때 exploration & exploitation 관점에서 LB는 exploration이 부족하다는 의미임

또한 여러 실험들을 통해 LB는 starting point에 가까운 minimizer를 향해 가는데 반면 SB는 멀리 이동함을 관측함
(LB methods tend to attracted to minimizers close to the starting point, whereas SB methods move away and locate minimizers that are farther away)

<img width="986" alt="image" src="https://user-images.githubusercontent.com/49643709/167088664-1e2e7383-2832-4ab1-b68a-5ed16efece9f.png">

LB의 경우 학습할수록 sharpness가 커지는데 반해 SB는 작아짐을 확인할 수 있음

### 4.1. Data Augmentation
<img width="862" alt="image" src="https://user-images.githubusercontent.com/49643709/167089585-8a56c0eb-c2fa-4431-8857-7cc4ddb392ce.png">
<img width="787" alt="image" src="https://user-images.githubusercontent.com/49643709/167089820-1ba196a1-6556-453d-82c1-8b26bb2f6ccc.png">

Loss function의 geometry를 바꿀 수 있는 방법은 objective function을 바꾸거나 training dataset을 바꾸면 됨

저자들은 aggressive data augmentation을 통해 이를 해결해보고자 했음
(horizontal reflections, random rotations up to 10 degree, random translation up to 0.2 times the size of the image)

LB method가 SB method와 comparable한 accuracy를 가질 수 있었고 sharpness도 감소했지만, sharpness of minima는 여전히 존재함

### 4.2. Conservative Training
<img width="1014" alt="image" src="https://user-images.githubusercontent.com/49643709/167090096-17c361c7-f788-4d5d-9198-53be37657ce9.png">
<img width="1016" alt="image" src="https://user-images.githubusercontent.com/49643709/167090182-871a6375-a954-4506-b41e-001e21c73524.png">
<img width="910" alt="image" src="https://user-images.githubusercontent.com/49643709/167090803-8e7acd03-3a8d-4cc5-9471-8e210cf40695.png">

기존의 weight update는 2번인데 (SGD), 다른 논문에서 제안한 방식인 5번으로 weight update

Test accuracy가 어느정도 증가했으나, sensitivity 문제를 해결하지 못함

### 4.3. Robust Training
<img width="733" alt="image" src="https://user-images.githubusercontent.com/49643709/167091087-ac0eb0f4-ff68-4686-8a53-de54a4a32efc.png">

Classical optimization은 valley의 가장 낮은 포인트를 찾는 반면 robust optimization은 epsilon-disc가 가장 낮은 포인트를 찾음

Intuitive appeal과 별개로 해당 방법과 다른 논문에서 제시되는 stability training 모두 generalization improvement가 없었음
(this strategy did not improve generalization, no generalization benefit from the stability training)

Robust training, stability training 모두 unmodified (baseline)과 유사한 test accuracy, sharpness, parametric plot을 보임
